File: services/inventory-service/internal/domain/events/inventory_events.go
package events

import "time"

// BaseEvent represents the common structure for all events
type BaseEvent struct {
	EventID       string  `json:"eventId"`
	EventType     string  `json:"eventType"`
	Timestamp     string  `json:"timestamp"`
	Version       string  `json:"version"`
	CorrelationID *string `json:"correlationId,omitempty"`
	Source        string  `json:"source"`
}

// StockReservedPayload contains the data for a stock reserved event
type StockReservedPayload struct {
	ReservationID string    `json:"reservationId"`
	ProductID     string    `json:"productId"`
	Quantity      int       `json:"quantity"`
	OrderID       string    `json:"orderId"`
	UserID        string    `json:"userId"`
	ExpiresAt     time.Time `json:"expiresAt"`
	ReservedAt    time.Time `json:"reservedAt"`
}

// StockReservedEvent represents a stock reservation event
type StockReservedEvent struct {
	BaseEvent
	Payload StockReservedPayload `json:"payload"`
}

// StockConfirmedPayload contains the data for a stock confirmed event
type StockConfirmedPayload struct {
	ReservationID string    `json:"reservationId"`
	ProductID     string    `json:"productId"`
	Quantity      int       `json:"quantity"`
	OrderID       string    `json:"orderId"`
	UserID        string    `json:"userId"`
	ConfirmedAt   time.Time `json:"confirmedAt"`
}

// StockConfirmedEvent represents a stock confirmation event
type StockConfirmedEvent struct {
	BaseEvent
	Payload StockConfirmedPayload `json:"payload"`
}

// StockReleasedPayload contains the data for a stock released event
type StockReleasedPayload struct {
	ReservationID string    `json:"reservationId"`
	ProductID     string    `json:"productId"`
	Quantity      int       `json:"quantity"`
	OrderID       string    `json:"orderId"`
	UserID        string    `json:"userId"`
	Reason        string    `json:"reason"` // "order_cancelled", "reservation_expired", "manual_release"
	ReleasedAt    time.Time `json:"releasedAt"`
}

// StockReleasedEvent represents a stock release event
type StockReleasedEvent struct {
	BaseEvent
	Payload StockReleasedPayload `json:"payload"`
}

// StockFailedPayload contains the data for a stock operation failure event
type StockFailedPayload struct {
	OperationType string    `json:"operationType"` // "reserve", "confirm", "release"
	ProductID     string    `json:"productId"`
	Quantity      *int      `json:"quantity,omitempty"`
	OrderID       string    `json:"orderId"`
	UserID        string    `json:"userId"`
	ReservationID *string   `json:"reservationId,omitempty"`
	ErrorCode     string    `json:"errorCode"`
	ErrorMessage  string    `json:"errorMessage"`
	FailedAt      time.Time `json:"failedAt"`
}

// StockFailedEvent represents a stock operation failure event
type StockFailedEvent struct {
	BaseEvent
	Payload StockFailedPayload `json:"payload"`
}

// Event routing keys
const (
	RoutingKeyStockReserved  = "inventory.stock.reserved"
	RoutingKeyStockConfirmed = "inventory.stock.confirmed"
	RoutingKeyStockReleased  = "inventory.stock.released"
	RoutingKeyStockFailed    = "inventory.stock.failed"
)

// Exchange name
const ExchangeInventoryEvents = "inventory.events"

// Event source
const SourceInventoryService = "inventory-service"

// Event version
const EventVersion = "1.0.0"

File: shared/types/src/events/inventory.events.ts
import { z } from 'zod';

/**
 * Base Event Schema
 * All events extend this base structure for consistency
 */
export const BaseEventSchema = z.object({
  eventId: z.string().uuid().describe('Unique identifier for this event'),
  eventType: z.string().describe('Type of event (routing key)'),
  timestamp: z.string().datetime().describe('ISO 8601 timestamp when event occurred'),
  version: z.string().default('1.0.0').describe('Event schema version'),
  correlationId: z.string().uuid().optional().describe('ID to correlate related events'),
  source: z.string().describe('Service that emitted the event'),
});

export type BaseEvent = z.infer<typeof BaseEventSchema>;

/**
 * Stock Reserved Event
 * Emitted by Inventory Service when stock is successfully reserved
 */
export const StockReservedEventSchema = BaseEventSchema.extend({
  eventType: z.literal('inventory.stock.reserved'),
  source: z.literal('inventory-service'),
  payload: z.object({
    reservationId: z.string().uuid().describe('Unique reservation identifier'),
    productId: z.string().describe('Product identifier'),
    quantity: z.number().int().positive().describe('Quantity reserved'),
    orderId: z.string().uuid().describe('Order that triggered the reservation'),
    userId: z.string().uuid().describe('User who made the reservation'),
    expiresAt: z.string().datetime().describe('When the reservation expires if not confirmed'),
    reservedAt: z.string().datetime().describe('When the reservation was made'),
  }),
});

export type StockReservedEvent = z.infer<typeof StockReservedEventSchema>;

/**
 * Stock Confirmed Event
 * Emitted by Inventory Service when a reservation is confirmed
 */
export const StockConfirmedEventSchema = BaseEventSchema.extend({
  eventType: z.literal('inventory.stock.confirmed'),
  source: z.literal('inventory-service'),
  payload: z.object({
    reservationId: z.string().uuid().describe('Reservation identifier that was confirmed'),
    productId: z.string().describe('Product identifier'),
    quantity: z.number().int().positive().describe('Quantity confirmed'),
    orderId: z.string().uuid().describe('Order that was confirmed'),
    userId: z.string().uuid().describe('User who owns the order'),
    confirmedAt: z.string().datetime().describe('When the confirmation occurred'),
  }),
});

export type StockConfirmedEvent = z.infer<typeof StockConfirmedEventSchema>;

/**
 * Stock Released Event
 * Emitted by Inventory Service when a reservation is released/cancelled
 */
export const StockReleasedEventSchema = BaseEventSchema.extend({
  eventType: z.literal('inventory.stock.released'),
  source: z.literal('inventory-service'),
  payload: z.object({
    reservationId: z.string().uuid().describe('Reservation identifier that was released'),
    productId: z.string().describe('Product identifier'),
    quantity: z.number().int().positive().describe('Quantity released back to stock'),
    orderId: z.string().uuid().describe('Order that triggered the release'),
    userId: z.string().uuid().describe('User who owns the order'),
    reason: z.enum(['order_cancelled', 'reservation_expired', 'manual_release']).describe('Why the stock was released'),
    releasedAt: z.string().datetime().describe('When the release occurred'),
  }),
});

export type StockReleasedEvent = z.infer<typeof StockReleasedEventSchema>;

/**
 * Stock Failed Event
 * Emitted by Inventory Service when a stock operation fails
 */
export const StockFailedEventSchema = BaseEventSchema.extend({
  eventType: z.literal('inventory.stock.failed'),
  source: z.literal('inventory-service'),
  payload: z.object({
    operationType: z.enum(['reserve', 'confirm', 'release']).describe('Which operation failed'),
    productId: z.string().describe('Product identifier'),
    quantity: z.number().int().positive().optional().describe('Quantity involved in the operation'),
    orderId: z.string().uuid().describe('Order that triggered the operation'),
    userId: z.string().uuid().describe('User who owns the order'),
    reservationId: z.string().uuid().optional().describe('Reservation ID if applicable'),
    errorCode: z.string().describe('Error code for troubleshooting'),
    errorMessage: z.string().describe('Human-readable error message'),
    failedAt: z.string().datetime().describe('When the failure occurred'),
  }),
});

export type StockFailedEvent = z.infer<typeof StockFailedEventSchema>;

/**
 * Stock Depleted Event
 * Emitted by Inventory Service when available stock reaches zero
 */
export const StockDepletedEventSchema = BaseEventSchema.extend({
  eventType: z.literal('inventory.stock.depleted'),
  source: z.literal('inventory-service'),
  payload: z.object({
    productId: z.string().describe('Product identifier that is now depleted'),
    orderId: z.string().uuid().describe('Order that caused the depletion'),
    userId: z.string().uuid().describe('User who owns the order'),
    depletedAt: z.string().datetime().describe('When the stock was depleted'),
    lastQuantity: z.number().int().positive().describe('The last quantity that was reserved/confirmed before depletion'),
  }),
});

export type StockDepletedEvent = z.infer<typeof StockDepletedEventSchema>;

/**
 * Union type of all inventory events
 */
export const InventoryEventSchema = z.discriminatedUnion('eventType', [
  StockReservedEventSchema,
  StockConfirmedEventSchema,
  StockReleasedEventSchema,
  StockFailedEventSchema,
  StockDepletedEventSchema,
]);

export type InventoryEvent = z.infer<typeof InventoryEventSchema>;

/**
 * Validation helper function
 */
export function validateInventoryEvent(data: unknown): InventoryEvent {
  return InventoryEventSchema.parse(data);
}

/**
 * Safe validation that returns errors instead of throwing
 */
export function safeValidateInventoryEvent(
  data: unknown
): { success: true; data: InventoryEvent } | { success: false; error: z.ZodError } {
  const result = InventoryEventSchema.safeParse(data);
  if (result.success) {
    return { success: true, data: result.data };
  }
  return { success: false, error: result.error };
}
